src/storage/src/hummock/iterator/mod.rs

```diff
@@ -106,18 +118,168 @@ pub trait HummockIterator: Send + Sync {
     /// - Do not decide whether the position is valid or not by checking the returned error of this
     ///   function. This function WON'T return an `Err` if invalid. You should check `is_valid`
     ///   before starting iteration.
-    async fn seek(&mut self, key: &[u8]) -> HummockResult<()>;
+    fn seek<'a>(&'a mut self, key: &'a [u8]) -> Self::SeekFuture<'a>;
 
     /// take local statistic info from iterator to report metrics.
     fn collect_local_statistic(&self, _stats: &mut StoreLocalStatistic) {}
 }
 
-pub trait ForwardHummockIterator = HummockIterator<Direction = Forward>;
-pub trait BackwardHummockIterator = HummockIterator<Direction = Backward>;
+/// This is a placeholder trait used in `HummockIteratorUnion`
+pub struct PhantomHummockIterator<D: HummockIteratorDirection> {
+    _phantom: PhantomData<D>,
+}
+
+impl<D: HummockIteratorDirection> HummockIterator for PhantomHummockIterator<D> {
+    type Direction = D;
+
+    type NextFuture<'a> = impl Future<Output = HummockResult<()>> + 'a;
+    type RewindFuture<'a> = impl Future<Output = HummockResult<()>> + 'a;
+    type SeekFuture<'a> = impl Future<Output = HummockResult<()>> + 'a;
+
+    fn next(&mut self) -> Self::NextFuture<'_> {
+        async { unreachable!() }
+    }
+
+    fn key(&self) -> &[u8] {
+        unreachable!()
+    }
+
+    fn value(&self) -> HummockValue<&[u8]> {
+        unreachable!()
+    }
+
+    fn is_valid(&self) -> bool {
+        unreachable!()
+    }
+
+    fn rewind(&mut self) -> Self::RewindFuture<'_> {
+        async { unreachable!() }
+    }
+
+    fn seek<'a>(&'a mut self, _key: &'a [u8]) -> Self::SeekFuture<'a> {
+        async { unreachable!() }
+    }
+}
+
+pub enum HummockIteratorUnion<
```

Comment:
I suggest adding more docs (like what you mentioned in the PR descriptions) in the codes to explain why we need HummockIteratorUnion and why we have 4 type in the enum so that people can understand the background without searching for this PR.

Reply:
+1 for this.